File: DoublyLinkedList.cpp
# include<iostream>
using namespace std;

class Node{
    public:
    int data;
    Node* next;
    Node* prev;
    
    Node(int val){
        data = val;
        next = NULL;
        prev = NULL;
    }
};

class List{
    public:
    int count;
    Node* head;
    List(){
        count = 0;
        head = NULL;
    }
};

void Insert_Node_at_start(List &list , Node* n){
    if(list.count == 0){
        list.head = n;
        list.count++;
    }
    else{
        Node* temp = list.head;
        n->next = list.head;
        list.head = n;
        temp->prev = n;
        list.count++;
    }
}

void Insert_Node_at_End(List &list , Node* n){
    if(list.count == 0){
        list.head = n;
        list.count++;
    }
    else{
        Node* temp = list.head;
        while(temp->next != NULL){
            temp = temp->next;
        }
        temp->next = n;
        n->prev = temp;
        list.count++;
    }
}

void Remove_Node_from_start(List &list){
    if(list.count == 0){
        cout << "List is empty" << endl;
    }
    else{
        Node* temp = list.head;
        list.head = temp->next;
        temp->next->prev = NULL;
        temp->next = NULL;
        list.count--;
    }
}

void Remove_Node_from_end(List &list){
    if(list.count == 0){
        cout << "List is empty" << endl;
    }
    else if(list.count == 1){
        list.head = NULL;
    }
    else{
        Node* temp = list.head;
        Node* prev = list.head;
        while(temp->next != NULL){
            prev = temp;
            temp = temp->next;
        }
        prev->next = NULL;
        temp->prev = NULL;
        list.count--;
    }
}

void Print_list(List &list){
    Node* temp = list.head;
    while(temp != NULL){
        cout << temp->data << endl;
        temp = temp->next;
    }
}

int main(){
    List My_list;
    Node* n1 = new Node(10);
    Node* n2 = new Node(20);
    Node* n3 = new Node(30);
    Node* n4 = new Node(40);
    Node* n5 = new Node(50);
    
    Insert_Node_at_start(My_list , n1);
    Insert_Node_at_End(My_list , n2);
    Insert_Node_at_End(My_list , n3);
    Insert_Node_at_End(My_list , n4);
    Insert_Node_at_End(My_list , n5);
    Remove_Node_from_end(My_list);
    Print_list(My_list);
    cout << My_list.count;
}

File: Doubly_Linked_List.py
class Node:
    def __init__(self , data):
        self.data = data
        self.next = None
        self.prev = None
        
class Doubly_Linked_List:
    def __init__(self):
        self.head = None
        
    def append(self , data):
        newNode = Node(data)
        if not self.head:
            temp = self.head
            self.head = newNode
            newNode.prev = temp
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        
        temp.next = newNode
        newNode.prev = temp
        
    def delete_Node(self , data):
        if self.head.data == data:
            temp = self.head
            self.head = self.head.next
            self.head.prev = temp
            return
        temp = self.head
        while temp:
            if temp.next == None:
                temp.prev.next = None
                temp.prev = None
                return
            if temp.data == data:
                temp.prev.next = temp.next
                temp.next.prev = temp.prev
                temp.next = None
                temp.prev = None
            temp = temp.next
        
    def print_List(self):
        temp = self.head
        while temp:
            print(temp.data)
            temp = temp.next
        print("END")
        
    def display_backwards(self):
        temp = self.head
        while temp.next:
            temp = temp.next
        while temp:
            print(temp.data)
            temp = temp.prev
        print("END")
        
        
My_List = Doubly_Linked_List()
My_List.append(10)
My_List.append(20)
My_List.append(30)
My_List.append(40)
My_List.append(50)
My_List.delete_Node(50)
My_List.print_List()


